src/DominantJuice.sol

```diff
@@ -0,0 +1,399 @@
+import {JBRedeemParamsData} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBRedeemParamsData.sol";
+import {JBRedemptionDelegateAllocation3_1_1} from
+    "@jbx-protocol/juice-contracts-v3/contracts/structs/JBRedemptionDelegateAllocation3_1_1.sol";
+import {JBTokenAmount} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBTokenAmount.sol";
+import {AccessControl} from "@openzeppelin/contracts/access/AccessControl.sol";
+import {UD60x18, ud, add, pow, powu, div, mul, wrap, unwrap} from "@prb/math/UD60x18.sol";
+
+/**
+ * @title Dominant Juice
+ * @author Armand Daigle
+ * @notice This contract is a JB Data Source, JB Pay Delegate, JB Redemption Delegate, and a Refund Bonus Escrow
+ * implementation. Campaigns, interchangeable with cycles from here on, that use this implementation as a Data Source
+ * will incorporate Alex Tabarrok's dominant assurance mechanism, which incentivize early pledgers with a campaign
+ * creator-provided refund bonus that can be withdrawn from this contract after a failed campaign.
+ * @dev This dominant assurance contract (DAC) uses the JB Data Source, Pay Delegate, and Pay Redemption interfaces and
+ * extends the functionality of a Juicebox project. Once a JB project is created, the constructor stores project data
+ * and the pertinent Juicebox architecture contracts. This contract is written for a single cycle campaign. If projects
+ * want dominant assurance on more cycles, refactoring is required.
+ */
+contract DominantJuice is
+    IJBFundingCycleDataSource3_1_1,
+    IJBPayDelegate3_1_1,
+    IJBRedemptionDelegate3_1_1,
+    AccessControl
+{
+    bytes32 private constant CAMPAIGN_MANAGER_ROLE = keccak256("CAMPAIGN_MANAGER_ROLE");
+
+    /// Juicebox Contracts
+    IJBController3_1 public immutable controller;
+    IJBDirectory public immutable directory; // directory of terminals and controllers for projects.
+    IJBFundAccessConstraintsStore public immutable fundAccessConstraintsStore;
+    IJBSingleTokenPaymentTerminalStore3_1_1 public immutable paymentTerminalStore;
+    IJBSingleTokenPaymentTerminal public immutable paymentTerminal;
+
+    /// Main Campaign Parameters
+    struct Campaign {
+        uint256 projectId;
+        uint256 cycleTarget;
+        uint256 cycleExpiryDate;
+        uint256 minimumPledgeAmount;
+        uint256 totalRefundBonus;
+    }
+
+    struct FundingStatus {
+        uint256 totalPledged;
+        uint256 percentOfGoal;
+        bool isTargetMet;
+        bool hasCycleExpired;
+        bool hasCreatorWithdrawnAllFunds;
+    }
+
+    struct Pledgers {
+        uint256 totalAmountPledged;
+        UD60x18 totalPledgeWeight;
+        mapping(address => UD60x18) pledgerWeight;
+        mapping(address => bool) hasBeenRefunded;
+    }
+
+    /// State Variables
+    Campaign private campaign;
+    Pledgers private pledgers;
+    UD60x18 private immutable rateOfDecay = ud(0.99e18);
```

Comment:
should be a constant if possible since it doesn't need to be storage.

Reply:
When I tried to make it constant last weekend, it was giving me this error: “Initial value for constant variable has to be compile-time constant.” That’s why I made it immutable storage. Do you know a way around that?

Reply:
ah, ok, no worries